### ISSUE TITLE
math: Modulus operator

### ISSUE THREAD

**Issue by matteodelabre** [sile-typesetter/sile#1494]
As far as I can tell, SILE’s `math` package doesn’t currently have support for the various forms of the modulus operator. I’d be happy to provide a patch, but I wanted to first discuss this here to make sure I’m not missing something.

Expected presentation
---------------------

The modulus has two main forms.

1. As an operation that refers to the remainder of a division, it is usually written in infix notation, using the `mod` or `%` signs, or in prefix notation. In LaTeX,

```tex
\documentclass{minimal}
\begin{document}
    \[42 \bmod 11 = 9\]
    \[42 \mathbin{\%} 11 = 9\]
    \[\mathrm{mod}(42, 11) = 9\]
\end{document}
```
renders as

![Screenshot from 2022-07-24 16-49-30](https://user-images.githubusercontent.com/1370040/180665338-3692eb99-f66c-4b40-8700-3647a6735409.png)

2. As part of a modular congruence relation, it is usually written at the end of the line, separated by a quad space, optionally within parentheses. In LaTeX,

```tex
\documentclass{minimal}
\usepackage{amsmath}
\begin{document}
    \[42 \equiv 9 \pmod{11}\]
    \[42 \equiv 9 \mod{11}\] % \mod provided by amsmath
\end{document}
```

renders as:

![Screenshot from 2022-07-24 16-48-32](https://user-images.githubusercontent.com/1370040/180665300-e5fc44e8-d190-4bb6-bf31-a7603f4fe1e4.png)

In SILE
-------

Trying to set the forms listed above in SILE leads to incorrect spacing, and the form that uses the `%` symbol cannot be written in `\math` since the `\%` command does not exist there.

```tex
\begin{document}
    \script[src=packages/math]
    \math[mode=display]{42 \mo{mod} 11 = 9}
    \mathml[mode=display]{\mrow{\mn{42}\mo{\%}\mn{11}\mo{=}\mn{9}}}
    \math[mode=display]{\mi{mod}(42, 11) = 9}
    \math[mode=display]{42 ≡ 9 (\mo{mod} 11)}
    \math[mode=display]{42 ≡ 9 \mo{mod} 11}
\end{document}
```

This renders as:

![Screenshot from 2022-07-24 16-55-11](https://user-images.githubusercontent.com/1370040/180665530-b6d17411-007b-40dd-b50e-f6cd2e0cbc5a.png)

Solutions
---------

If I understand correctly, the `\%` operator should be added to the definitions of the `packages/math/texlike.lua` file.

The spacing around the `mod` and `%` operators can fixed by declaring them as binary operators in the [symbolDefaults](https://github.com/sile-typesetter/sile/blob/c7db5d5b3c5c6756ffe7e923ea3b938148167b2b/packages/math/unicode-symbols.lua#L2550) table (is this the right way to do it?).

```tex
\begin{document}
    \script[src=packages/math]
    \script{
        local syms = require('packages.math.unicode-symbols')
        syms.symbolDefaults['%'] = { atomType = 2 }
        syms.symbolDefaults['mod'] = { atomType = 2 }
    }
    \math[mode=display]{42 \mo{mod} 11 = 9}
    \mathml[mode=display]{\mrow{\mn{42}\mo{\%}\mn{11}\mo{=}\mn{9}}}
    \math[mode=display]{\mi{mod}(42, 11) = 9}
    \math[mode=display]{42 ≡ 9 (\mo{mod} 11)}
    \math[mode=display]{42 ≡ 9 \mo{mod} 11}
\end{document}
```

I think it would be useful if the authors could override the “atom type” for a given symbol in their documents, for example if they need to define a custom operator unknown to SILE that should be treated as a binary operator. LaTeX provides `\mathord`, `\mathop`, `\mathbin`, `\mathrel`, …, for this purpose. Perhaps this could be handled as an option to the `\mo` command?

Finally, I could not find an easy way to add a quad space before `mod` in the 2nd form.

**Comment by alerque** [comment]
As far as I know you aren't missing anything, but I'm not deeply familiar with the math package internals. The primary contributor @OlivierNicole would know more, but last I heard we was pretty busy in other fields and not currently typesetting anything. Perhaps he'll drop some quick pointers if he gets a chance.

In the mean time if you're up for contributing proper commands to facilitate this I'd say go for it.

What context here would determine when to add or not add the quad-space?

**Comment by OlivierNicole** [comment]
> If I understand correctly, the `\%` operator should be added to the definitions of the `packages/math/texlike.lua` file.

Yes, it is indeed missing.

> The spacing around the `mod` and `%` operators can fixed by declaring them as binary operators in the [symbolDefaults](https://github.com/sile-typesetter/sile/blob/c7db5d5b3c5c6756ffe7e923ea3b938148167b2b/packages/math/unicode-symbols.lua#L2550) table (is this the right way to do it?).
> 
> ```
> \begin{document}
>     \script[src=packages/math]
>     \script{
>         local syms = require('packages.math.unicode-symbols')
>         syms.symbolDefaults['%'] = { atomType = 2 }
>         syms.symbolDefaults['mod'] = { atomType = 2 }
>     }
>     \math[mode=display]{42 \mo{mod} 11 = 9}
>     \mathml[mode=display]{\mrow{\mn{42}\mo{\%}\mn{11}\mo{=}\mn{9}}}
>     \math[mode=display]{\mi{mod}(42, 11) = 9}
>     \math[mode=display]{42 ≡ 9 (\mo{mod} 11)}
>     \math[mode=display]{42 ≡ 9 \mo{mod} 11}
> \end{document}
> ```

It's a sensible way to do it. It would probably be useful to have an `\operator` command to do it.

> I think it would be useful if the authors could override the “atom type” for a given symbol in their documents, for example if they need to define a custom operator unknown to SILE that should be treated as a binary operator. LaTeX provides `\mathord`, `\mathop`, `\mathbin`, `\mathrel`, …, for this purpose. Perhaps this could be handled as an option to the `\mo` command?

This is also the kind of things that can be handled by an `\operator` command. Now, regarding atom types and spacing around operators: atom types are a TeX convention that does not appear in [the MathML standard](https://www.w3.org/TR/MathML3), however we used it in SILE because it provides good default spacing in most cases.

This is not in conflict with MathML as MathML lets the renderer mostly free to draw whatever spaces it wants around operators. With one exception: the standard specifies three attributes for the `mo` element, `lspace`, `rspace`, and `form` (which is one of `prefix`, `infix`, `postfix`). A perfectly compliant MathML renderer, if it exists, should honor these attributes. Currently we ignore them.

All that being said, you should be able to do pretty much the same as `\mathord`, etc. in TeX by doing:
```tex
\mo[atomtype=rel]{mod}
```
This should *almost* work. The only missing thing is to convert the text “rel” into `atomType.relationalOperator`. Also, the operator initialization in `math/base-elements.lua` is wrong as it will always use the default settings for the operator (I'm noticing this just now):
https://github.com/sile-typesetter/sile/blob/c7db5d5b3c5c6756ffe7e923ea3b938148167b2b/packages/math/base-elements.lua#L791-L794
We should just remove this.

> Finally, I could not find an easy way to add a quad space before `mod` in the 2nd form.

You could just define a `\quad` math macro as
```tex
\def{quad}{\mspace[width=1em]}
```
for example. (Actually, the width of a quad should be equal to the current font size, but I'm not sure how to do that right now.)

`\quad` would certainly be a useful addition to the default macros :)

**Comment by matteodelabre** [comment]
Thank you both for the pointers! I will prepare a patch.

**Comment by OlivierNicole** [comment]
Feel free to ping me with questions.

**Comment by alerque** [comment]
> (Actually, the width of a quad should be equal to the current font size, but I'm not sure how to do that right now.)

The `em` unit is *defined as* the current font size, so you've already got it.

> `\quad` would certainly be a useful addition to the default macros :)

We already have a `\quad` and `\qquad` defined in the *plain* class, which most other classes inherit from. It is essentially just a shortcut `SILE.call("glue", { width = "1em" })`. I suppose it is just math mode that needs a similar function.

**Comment by matteodelabre** [comment]
It looks like `\def`s wrap their contents into `mrow`s, so the following:
```tex
\math{
  \def{bmod}{\mo{mod}}
  42 \bmod 11 = 9
}
```
is compiled into:
```tex
 \math{\mrow{\mn{42}\mrow{\mo{mod}}\mn{11}\mo{=}\mn{9}}}
```
(notice the extra `mrow` around `\mo{mod}`).

This behavior breaks spacing around the operator, as the wrapping `mrow` does not carry the `atomType` from its child (this is not specific to `mod`, it would happen with any other operator called from a macro).

To fix this, we could change the spacing rules lookup logic to recurse into any adjacent `mrow`s and find the first child with a non-nil `atomType`. But in some circumstances it may be useful to forcibly disable spacing around parts of a formula by manually adding `mrow`s. In TeX, adding a group around an operator (e.g., `1{+}1{=}2`) _does_ disable spacing, which suggests it does not implement this strategy.

I believe the correct fix would instead be to change the macro behavior to not add any wrapping tags. This would also be more in line with the expected behavior of macros, as they are supposed to simply expand to their contents. This should be fairly easy to implement, so I can include it in my patch.

Any thoughts?

**Comment by alerque** [comment]
That sounds reasonable to me (again as a math outsider). I suppose we should consider whether to label that a fix or a breaking change...

**Comment by OlivierNicole** [comment]
I guess it would make sense.

As an aside, I'm reading that a compliant MathML renderer is supposed to treat a single-element `mrow` exactly in the same way as the contained element (which could be achieved by a simplification pass): https://www.w3.org/TR/MathML3/chapter3.html#id.3.3.1.3.1. But producing no `mrow` around the result of macro expansion can still make sense even outside of this particular case.

**Comment by matteodelabre** [comment]
> I suppose we should consider whether to label that a fix or a breaking change...

It does fit in both. Implementing this has changed the rendering on the `math-macros` regression test, bringing the subscript closer to the base (old in red, new in green). The new spacing looks like the correct one.

![math-macros-diff](https://user-images.githubusercontent.com/1370040/181409503-3e08ff5b-5c5b-468b-83f2-e93888261454.png

> As an aside, I'm reading that a compliant MathML renderer is supposed to treat a single-element `mrow` exactly in the same way as the contained element (which could be achieved by a simplification pass)

Interesting, thanks! So the `mrow` element doesn’t quite map to groups in TeX.

**Comment by alerque** [comment]
If the new spacing can be considered correct or more correct and nothing that would have been considered okay before is seriously worse without redoing markup I think we can call this a fix/feature not a breaking change. We don't guarantee no document reflows between versions, only that documents compile and do an the expected thing in a given scenario. We'll look at the test output based on your changes and if it looks correct we can update the expected test output.

**Comment by alerque** [comment]
This was implemented in #1496 but I'm waiting to close this issue as a way to track that documentation has not yet been updated to match.
